api/turing/service/router_service.go

```diff
@@ -112,21 +116,28 @@ func (service *routersService) Delete(router *models.Router) error {
 		return errors.New("router must have valid primary key to be deleted")
 	}
 	tx := service.db.Begin()
-	var routerVersions []*models.RouterVersion
 
 	// remove associations
 	router.ClearCurrRouterVersion()
 	tx.Save(router)
 
-	service.db.Model(router).Related(&routerVersions)
+	var routerVersions []*models.RouterVersion
+	if err := service.db.
```

Comment:
1. Side note: we are on gorm v1, so the correct version of docs is here: https://v1.gorm.io/docs/preload.html
This is the first time I hear about `when it is accessed` and I'm not even sure if it's possible to easily implement such pattern in golang. (I've seen it before in Hibernate with Java, but that's a different story). So it's not a performance optimization, but rather an instruction for gorm to either load the assosiated entry(ies) or not. 
With that said, if preloading doesn't happen, then corresponding `routerVersion.Enricher` and `routerVersion.Ensembler` are always `nil`, hence – not removed from the database. So your assumption leads you by the wrong path.
> And if so, is it not enough to always delete the router version only after the enricher/ensembler is deleted? 
This is wrong too. `router_versions` table references `enrichers.id`/`ensemblers.id`, so you can't delete Enricher/Ensembler first, because in this case the record in `router_versions` table will reference non-existing records in `enrichers`/`ensemblers` tables, which is prohibited. 

But now, I realized that we don't really need to `Preload` `Enricher`/`Ensembler` and we can just use `EnsemblerId`/`EnricherId` and delete corresponding records by their primary keys. So, thanks for bringing my attention to this :)

2. Hooks in general are harder to navigate and understand, so I'd prefer the (current) explicit deletion method.

Reply:
Okay, my assumption that the support for eager loading in Go Gorm should mean there is an implicit lazy loading comes from Django's ORM. I tried looking up docs for Go Gorm and verifying it locally - you are right that there is no "when it is accessed" and rather, it must at least be done with [Related](https://pkg.go.dev/github.com/jinzhu/gorm#DB.Related) at query time.

Thanks for the clarification!